### ISSUE TITLE
Dependency deprecation, futures

### ISSUE THREAD

**Issue by rgbkrk** [plotly/react-ipython-notebook#6]
Heyo!

The current `package.json` won't install because of changes to a package name:

```
npm ERR! addLocal Could not install /Users/kyle6475/code/src/github.com/plotly/react-ipython-notebook/lib
npm WARN deprecated mdast-react@0.3.0: Renamed to `remark-react`
npm ERR! Darwin 14.5.0
npm ERR! argv "/Users/kyle6475/.nvm/versions/node/v5.1.0/bin/node" "/Users/kyle6475/.nvm/versions/node/v5.1.0/bin/npm" "install"
npm ERR! node v5.1.0
npm ERR! npm  v3.3.12

npm ERR! No version provided in package.json
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/kyle6475/code/src/github.com/plotly/react-ipython-notebook/npm-debug.log
```

*However*, I've got some React components for you that do translation of the notebook to the same format you would expect in the notebook (styling is up to you still). The one you're likely most interested in is [react-jupyter-display-area](https://github.com/nteract/react-jupyter-display-area).

Since this is being used in a live runnable version of the notebook, called [composition](https://github.com/nteract/composition), we require the props to be immutable, particularly Immutable.js.

```jsx
import Display from 'react-jupyter-display-area';

<Display outputs={this.props.outputs} />
```

We've also made a library for converting the on-disk notebook format to the one that tends to be used in memory, called [`commutable`](https://github.com/nteract/commutable).

```es6
var commutable = require('commutable');
commutable.fromJS(notebookJSON);
```

I've been planning on using this anyways for a static version of the notebook, so I thought I'd highlight this here since you're well on the way. Of note though, if you plan to host arbitrary notebooks be mindful of the fact that letting the user's raw HTML end up on a subdomain that has auth. There are solutions for that too and we can talk about that later.

**Comment by rgbkrk** [comment]
By the way, this will close out a few of the issues that are in the README here, namely that the display area react component will handle inline images.

**Comment by rgbkrk** [comment]
As for including MathJax or KaTeX, either one is easy enough to add to the head of the document.

**Comment by rreusser** [comment]
Oh wow, interesting… Yeah, the biggest blocker I was running into was getting markdown with equations to render in a react-friendly way. `mdast` was going through a major update before I left for the holidays, but if there's already a solution (`react-jupyter-display-area`), then that'd be way better and less ad-hoc. Feel free to steamroll the dependency here. Even if we did use `mdast`, it'd have to get ripped out and replaced with an updated anyway.

**Comment by wooorm** [comment]
Hey, author of mdast/remark chiming in, regarding the name change, it’s just that, plus a lot of great fixes :wink: Now, the whole MathJax/KaTeX isn’t there though, so if this package fixes that you should definitely refactor to that!

**Comment by rreusser** [comment]
Haha, thanks. Didn't mean to assume it was a bigger change than it was! My inclination was to intercept equations in the markdown rendering step, but perhaps the best solution is to let a markdown renderer do its thing and then just use default mathjax/katex after the fact to massage the content. I think the first step is to sync up on exactly what the business requirements are (like is setting markdown content with `dangerouslySetInnerHTML` acceptable? If these are coming from python notebooks, do we need 100% support for their equation syntax? (vs. customized delimiters or non-standard, i.e. katex, rendering)). Not a big deal, just hesitant to go down a path without understanding the target.

Thanks for chiming in though! It sounds like @rgbkrk maybe has a pretty solid approach already somewhat defined.

**Comment by rgbkrk** [comment]
Ah, cool. I've been using https://github.com/rexxars/react-markdown, though I'm not beholden to any. That one returns pure React elements as well. I'd actually love if we could rely on a math extension to CommonMark; [it hasn't been fully specced yet](http://talk.commonmark.org/t/mathematics-extension/457).

**Comment by rgbkrk** [comment]
I think the way I intro'ed the issue probably wasn't so great, since most of what I've concerned with is handling the notebook document as a whole. The markdown bits were quick adds, both for markdown cells and markdown in outputs.

**Comment by rreusser** [comment]
Okay, backing up a step just to clarify, the problem is this:

1. iPython notebooks allows markdown that includes inline and display math with delimiters (`$`/`$$` or maybe `\(...\)`? This is based on Pandoc, I think?)
2. As soon as you run it through a markdown parser, all that stuff gets mangled.
3. Your equations are useless

The solutions are:

1. Use a markdown parser that accounts for this and handles it magically (commonmark? is that a standard with this in mind; just not spec'd/implemented yet?)
2. Intercept equations *before* parsing the markdown and manage to pass it through unaffected. This was the logic in pulling out the magic function from KaTeX that locates equations. Then we massage the input—maybe wrap it in a code block so it's unaffected by markdown parsing, intercept it on the other side, and send it through MathJax/KaTeX.

Is this indeed the challenge? Does react-jupyter-display-area solve this problem?

**Comment by rgbkrk** [comment]
People are aware of it, it's just not implemented. When we relied on `marked`, we configured it to not mangle maths.

**Comment by rgbkrk** [comment]
> Does react-jupyter-display-area solve this problem?

Nope. It handles all the mimetypes that are missing.

As for Markdown cells themselves, those are inside of a class in Composition that haven't been extracted like `react-jupyter-display-area`.

**Comment by wooorm** [comment]
OK, I didn’t follow everything in this thread, but if you have some actionable things you’d want from a markdown compiler I may be able to advise on what’s possible re remark.

Am I reading correctly that you don’t want stuff like emphasis and strong emphasis to be rendered inside values delimited by `$`, `$$`, and `()`?

**Comment by rgbkrk** [comment]
Cool, I'm going to take remark for a spin in favor of the current markdown react library we were working with.

We'd like to not render anything between matching pairs of `$$`, `$`, or `\(` followed by `\)` (literal `\` there). In fact, if the parser can pick those up as entities which we can pass to another React component (likely KaTeX for some, additional styling for others).

**Comment by rgbkrk** [comment]
Probably worth noting that those tokens don't have to be escaped in inline code or code blocks (should let people write code as is).

**Comment by wooorm** [comment]
How about like this? [markdown-it-math](https://www.npmjs.com/package/react-rich-markdown). (I’d vote against supporting the latex-style markers as they do though)

**Comment by rgbkrk** [comment]
Certainly similar to that. The main reason I wouldn't use that is that [they're relying on dangerouslySetInnerHTML](https://github.com/mattecapu/react-rich-markdown/blob/master/source/Markdown.jsx#L54) instead of creating pure React components:

```jsx
<div className="markdown" dangerouslySetInnerHTML={{__html: this.Markdown.render(this.props.source)}}></div>
```

I think I would rather use [`remark-react`](https://github.com/mapbox/remark-react) or your remark-vdom package if that's easy enough to use from React.

**Comment by wooorm** [comment]
I’m currently working on a highlight.js-syntaxes compatible syntax highlighter producing a virtual dom (called [HAST](https://github.com/wooorm/hast), which should be easy to transform to react). I’d definitely be interested in creating the same for ascii2mathml (which looks [decoupled](https://github.com/runarberg/ascii2mathml/blob/master/src/parser.js), which is great). That would enable math to work for react/vdom/hast.

However, I’m guessing re-implementing KaTeX would be a much greater feat, quite a lot of hard work. But I’m not very familiar with it.

**Comment by rgbkrk** [comment]
There's always the option to leave any `$$...$$` unescaped, which we can let MathJax operate on globally (that's how it's done on http://math.stackexchange.com/ for instance). Re-implementing KaTeX would be hard, as is reimplementing LaTeX in general. The way that Khan Academy has been using KaTeX is in a bit of a dual mode here: https://github.com/Khan/react-components/blob/master/js/tex.jsx

**Comment by rreusser** [comment]
The source of uncertainty for me was the degree to which it needs to meet some standard so that existing notebooks aren't garbled when uploaded. (Or more to the point, whether some standard exists.) But if this work defines what is permissible, that opens things up a bit.

Tricky examples I've encountered:

- Can formulas span multiple lines? This might interfere with the order of operations in markdown parsing vs math parsing.
- From [math in markdown](https://github.com/cben/mathdown/wiki/math-in-markdown#pandoc): "Smart enough to parse nested $y = x^2 \hbox{ when $x > 2$}$ as single formula."

I love that you guys are working on this and you clearly have the general problem under control and I can't wait to see the result and I have my hands generally full at the moment, so I'll try to avoid further interference! I was worried from the beginning that I was maybe a bit out in left field, but if the focus and requirements are clearer than they were to me, I'm still glad to pitch in!

**Comment by wooorm** [comment]
Okay, so the approach is as follows, I think:

* [ ] Add `remark-math-block`, which parses `$$ ... $$` things into a `math` node;
* [ ] Add `remark-ascii2mathml`, which reads `math` nodes, transforms the math into a HAST tree, which patches that onto the node in such a way that `remark-html` can compile it to HTML.

The first has precedence. The second can come later.

To get the first going, I need to see several fixtures of what should be seen as math blocks, I’m not entirely sure yet what the syntax should be.

**Comment by rreusser** [comment]
Makes perfect sense to me. This is just a shot in the dark, but I made [a repo with some very very basic cases](https://github.com/rreusser/markdown-math-fixtures). I'm not sure html output is the target here, so it's really just five minutes toward maybe cementing precisely what we would or would not expect to be interpreted.

**Comment by wooorm** [comment]
Thanks. Am I right in assuming [this](https://raw.githubusercontent.com/rreusser/markdown-math-fixtures/master/fixtures/challenging-case.md) fixture works because the nested dollars are inside braces?

**Comment by rreusser** [comment]
Correct. It's unclear to me which of `\(`, `$`, and `$$`, should be accepted. That case may support leaving out `$`, or at the very least that a naive regex-based latex-syntax-ignorant solution may not be adequate.

**Comment by rgbkrk** [comment]
`\(` is not accepted.

Try it out on tmpnb.org

**Comment by rreusser** [comment]
Works for me. Branched fixtures into acceptable vs unacceptable.

**Comment by rgbkrk** [comment]
Awesome breaking it down into fixtures.

**Comment by rreusser** [comment]
Beyond that, is the goal to run mathjax or katex after remark has done its thing and all of this is moot? Or is the goal to intercept before remark does its thing and properly represent this in react—perhaps without reimplementing katex but without using `dangerouslySetInnerHTML` either.

**Comment by rreusser** [comment]
And I guess more to the point, @wooorm, yeah, that example works because the parser is presumably smart. :smile: I'm not actually aware of how remark parses input, but my first thought would be for remark to integrate that magic katex contrib function that does this job (perhaps? presumably? hopefully?) well already. I'm not sure how remark tokenizes and breaks up lines, so this may be nontrivial, but ideally it would offload the tricky job of using delimiters to locate equations onto what katex already does.

**Comment by rgbkrk** [comment]
Easy mode is to run Mathjax globally on the page (letting it convert plain `$$\LaTeX$$` over to images). They at least have a safe mode to make sure that embedded HTML won't run. We can probably rely on KaTeX for the time being, they have a similar mode.

**Comment by rreusser** [comment]
Yeah, makes sense. Then is the sole responsibility of `remark-math-block` to prevent equations from getting mangled by remark so that they're picked up and rendered by a mathjax/katex refresh after react rendering?

**Comment by rgbkrk** [comment]
Yeah. I'd prefer that we were able to use the math blocks ourselves directly, though letting mathjax run on the page does work.

**Comment by wooorm** [comment]
Darn, I cannot find that `auto-render` functionality as an npm package. Is that because it’s not published there, or am I missing something?

**Comment by rreusser** [comment]
Here's everything I know:

https://github.com/Khan/KaTeX/blob/master/contrib/auto-render/README.md

Appears not published individually, but I think you could use:

```javascript
require('katex/contrib/auto-render/splitAtDelimiters')
```

I think it says they pulled it from somewhere else, so you could also fork it locally or maybe even just make an npm package for it… (I definitely think the latter might be worthwhile).

**Comment by wooorm** [comment]
The `contrib` directory isn’t published :frowning:

**Comment by rreusser** [comment]
I can have a module up in like five minutes if that helps? I believe licensing permits it but need to double-check.

**Comment by wooorm** [comment]
I can do that, don’t worry :)

**Comment by rreusser** [comment]
All good. I have it 90% done. Can bail or continue.

**Comment by wooorm** [comment]
Wowie. Euhm. What API does your module expose? What I need is a function which expects a value in which the value may or may not start at character `0`: https://github.com/wooorm/remark/blob/master/doc/remarkplugin.3.md#function-tokenizereat-value-silent

**Comment by rreusser** [comment]
Sorry, I mean I just threw that file onto github as an npm module… this is something I've wanted too…

https://github.com/rreusser/split-at-math-delimiters

**Comment by wooorm** [comment]
(I’m making the remark interface, but that link shows the available info)

**Comment by wooorm** [comment]
Oh that’s perfect, the `startIndex` parameter is exactly what I needed!

**Comment by rreusser** [comment]
Gave you push access. Can think about it for a few minutes before publishing and npm install from github in the mean time.

**Comment by wooorm** [comment]
I’d say add an API section in the readme, other than that, I’m fine with publishing!

**Comment by rreusser** [comment]
Ahhh… `split([{type: 'text', data: 'test $y = 7$ and some more text'}], '$', '$', false)`

**Comment by wooorm** [comment]
Ok, weird, but thanks. :wink: 

Another question: `$` is a marker for an “inline” math, right? and `$$` for a block math. That’s pretty weird, because they will both occur inside a “paragraph” element ([this block example](https://raw.githubusercontent.com/rreusser/markdown-math-fixtures/master/fixtures/acceptable/double-dollar-sign.md)). Why then should, when compiling to HTML, that block math be wrapped in [a new paragraph](https://raw.githubusercontent.com/rreusser/markdown-math-fixtures/master/fixtures/acceptable/double-dollar-sign.html)?

**Comment by rreusser** [comment]
I think their code tries to handle the nesting of braces, but that doesn't mean it makes sense on every front. I have to get back to work at the moment, but you have push access, so if you need changes, don't hesitate! Or if you just want to pull the code into another project, feel free to disregard the repo.

**Comment by wooorm** [comment]
:+1:

**Comment by wooorm** [comment]
I’ve got a first version working.

Some question:

* Are triple dollars allowed? `$$$ foo $$ bar $$$`? Inline code in markdown works that way, which makes it easy to include any number of ticks / dollars inside spans.
* If not, is there any way to include more than one dollar inside a single-dollar fenced math, or two adjacent dollars inside double-fenced math spans? Note that non-CommonMark markdown does not allow escaped dollar characters: \$.
* Why do double-dollar fences math spans result in a paragraph break, *with* a [span inside](https://raw.githubusercontent.com/rreusser/markdown-math-fixtures/master/fixtures/acceptable/double-dollar-sign.html)?

**Comment by rreusser** [comment]
Regarding the third, take it with a grain of salt. That was just me guess what output might look like, which of course assumes that html output is even meaningful.

**Comment by wooorm** [comment]
Ah, okay, didn’t get that but thanks for pointing it out :)
